File: scripts/hooks/check_skill_references.py
#!/usr/bin/env python3
"""UserPromptSubmit hook: backstop for the Capability Existence Check rule.

Scans the submitted prompt for /slash-command style references and, for any that
match a skill or command actually installed on disk, injects a [CAPABILITY EXISTS]
notice into the model's context. This makes "that skill doesn't exist" a harder
mistake to make.

Contract: reads hook JSON on stdin, writes hook JSON on stdout, always exits 0.
Never blocks a prompt - a broken guard must not break the session.
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path

CLAUDE_DIR = Path.home() / ".claude"

# A slash reference at a word boundary: /skill-name, not a/path/fragment.
SLASH_REF = re.compile(r"(?:^|[\s(\[`\"'])/([a-zA-Z][a-zA-Z0-9_-]{1,63})")


def installed_capabilities() -> dict[str, str]:
    """Map capability name -> where it lives. Directory names are authoritative."""
    found: dict[str, str] = {}

    for root in (CLAUDE_DIR / "skills", CLAUDE_DIR / "plugins"):
        if not root.is_dir():
            continue
        for skill_md in root.rglob("SKILL.md"):
            found.setdefault(skill_md.parent.name, str(skill_md.parent))

    commands = CLAUDE_DIR / "commands"
    if commands.is_dir():
        for cmd in commands.rglob("*.md"):
            found.setdefault(cmd.stem, str(cmd))

    return found


def main() -> int:
    raw = sys.stdin.read()
    try:
        payload = json.loads(raw) if raw.strip() else {}
    except (ValueError, TypeError):
        payload = {}

    prompt = payload.get("prompt")
    if not isinstance(prompt, str) or not prompt:
        prompt = raw  # fall back to scanning the whole payload

    referenced = {m.group(1) for m in SLASH_REF.finditer(prompt)}
    if not referenced:
        return 0

    installed = installed_capabilities()
    hits = sorted(name for name in referenced if name in installed)
    if not hits:
        return 0

    lines = [f"  /{name}  ->  {installed[name]}" for name in hits]
    context = (
        "[CAPABILITY EXISTS] The prompt references capabilities that ARE "
        "installed on disk. Do not tell the user they don't exist, and do not "
        "substitute a different approach without saying so first:\n"
        + "\n".join(lines)
    )

    json.dump(
        {
            "hookSpecificOutput": {
                "hookEventName": "UserPromptSubmit",
                "additionalContext": context,
            },
            "suppressOutput": True,
        },
        sys.stdout,
    )
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except Exception:  # noqa: BLE001 - a guard must never break the session
        sys.exit(0)

File: scripts/hooks/check-skill-references.js
#!/usr/bin/env node
/**
 * UserPromptSubmit hook: capability existence check.
 *
 * Scans the user prompt for /slash-command references. For each token that
 * resolves to an installed skill directory (~/.claude/skills/<name>/ or a
 * project .claude/skills/<name>/), injects a reminder that the skill EXISTS,
 * so the model cannot claim it is unavailable nor silently substitute its own
 * plan.
 *
 * Match on DIRECTORY existence, not the manifest filename — gstack skills vary
 * the case (project-scaffold/skill.md vs office-hours/SKILL.md), so a
 * SKILL.md-based check produces false negatives.
 *
 * Fail-open: any error exits 0 with no output. Fires only on real on-disk
 * matches, so it can never produce a false "this exists" signal.
 */
const fs = require('fs');
const os = require('os');
const path = require('path');

function skillDirs() {
  const cfg = process.env.CLAUDE_CONFIG_DIR || path.join(os.homedir(), '.claude');
  const dirs = [path.join(cfg, 'skills')];
  if (process.env.CLAUDE_PROJECT_DIR) {
    dirs.push(path.join(process.env.CLAUDE_PROJECT_DIR, '.claude', 'skills'));
  }
  try { dirs.push(path.join(process.cwd(), '.claude', 'skills')); } catch (e) {}
  return [...new Set(dirs)];
}

function resolveSkill(token, dirs) {
  // Strip a plugin namespace prefix (plugin:skill -> skill).
  const name = token.includes(':') ? token.split(':').pop() : token;
  if (!name) return null;
  for (const d of dirs) {
    try {
      const p = path.join(d, name);
      if (fs.existsSync(p) && fs.statSync(p).isDirectory()) return p;
    } catch (e) {}
  }
  return null;
}

function main() {
  let input = '';
  process.stdin.setEncoding('utf8');
  process.stdin.on('data', c => { input += c; });
  process.stdin.on('end', () => {
    try {
      const data = JSON.parse(input);
      const prompt = (data.prompt || data.user_prompt || '').toString();

      // /token references: a slash at a word boundary, then a skill-ish name.
      // Leading [\s(] or start-of-string avoids matching paths like src/foo or and/or.
      const tokens = new Set();
      const re = /(?:^|[\s(])\/([a-zA-Z][a-zA-Z0-9_:-]*)/g;
      let m;
      while ((m = re.exec(prompt)) !== null) tokens.add(m[1]);
      if (tokens.size === 0) { process.exit(0); }

      const dirs = skillDirs();
      const found = [];
      for (const t of tokens) {
        const p = resolveSkill(t, dirs);
        if (p) found.push({ token: t, path: p });
      }
      if (found.length === 0) { process.exit(0); }

      const reminder = [
        '',
        '[CAPABILITY EXISTS — DO NOT CLAIM OTHERWISE]',
        'The user referenced these skills, which ARE installed on disk:',
        ...found.map(f => `  /${f.token}  ->  ${f.path}`),
        'Do NOT tell the user any of these is unavailable, does not exist, or cannot be found.',
        'Invoke it via the Skill tool, or ask one scoping question. Never substitute your own',
        'plan and present it as the only option.',
        ''
      ].join('\n');

      process.stdout.write(JSON.stringify({
        hookSpecificOutput: {
          hookEventName: 'UserPromptSubmit',
          additionalContext: reminder
        }
      }));
      process.exit(0);
    } catch (e) {
      process.exit(0);  // never block on hook errors
    }
  });
}

main();
